### PR TITLE
Migrate to hardhat-toolbox

### LINKED ISSUE #89: Add subscription ID to network config
Since VRF Subscription will be different chain to chain, it shouldn't be an environment variable, instead, in the config.

### LINKED ISSUE #112: Migrate from hardhat-waffle
Hardhat's latest [release](https://github.com/NomicFoundation/hardhat/releases/tag/hardhat@2.10.0) suggests migrating away from `hardhat-waffle` to `hardhat-toolbox`

### LINKED ISSUE #115: @ledgerheq/hw-transport-node-hid version ^4.3.0 fails on build
## Dependency `@ledgerheq/hw-transport-node-hid fails` on `^4.3.0` version with latest versions of yarn, node, and node-gyp.
```
make: *** [HID-hidraw.target.mk:120: Release/obj.target/HID-hidraw/src/HID.o] Error 1
make: Leaving directory '/tmp/test/node_modules/node-hid/build'
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/lib/node_modules/node-gyp/lib/build.js:201:23)
gyp ERR! stack     at ChildProcess.emit (node:events:527:28)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (node:internal/child_process:291:12)
gyp ERR! System Linux 5.18.15-arch1-1
gyp ERR! command "/usr/bin/node" "/usr/lib/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /tmp/test/node_modules/node-hid
gyp ERR! node -v v16.16.0
gyp ERR! node-gyp -v v9.1.0
gyp ERR! not ok

```

# Steps to reproduce
```
↪ node --version                                                                                                                                                              12:11:03
v16.16.0
matt at delleite in /t/test 
↪ yarn --version                                                                                                                                                              12:25:50
1.22.19
matt at delleite in /t/test 
↪ node-gyp --version                                                                                                                                                          12:25:54
v9.1.0
↪ mkdir node-hid-breaks
↪ cd node-hid-breaks
↪ yarn add "@ledgerhq/hw-transport-node-hid@^4.3.0"
```

A hotfix can be replacing line `158` of `yarn.lock` from
`"@ledgerhq/hw-transport-node-hid" "^4.3.0"`
to
`"@ledgerhq/hw-transport-node-hid" "^5.26.0"`
which is the next version seen in the rest of the packages.

Alternatively, you could remove hardware wallets dependencies altogether (or just ledgers', they may seem unnecessary depending on your situation).

I'll be submitting this minor modification as a pull request momentarily.

### LINKED ISSUE #136: Error running 'yarn' command
warning Pattern ["ethereumjs-abi@git+https://github.com/ethereumjs/ethereumjs-abi.git"] is trying to unpack in the same destination "/home/zoidburgh/.cache/yarn/v6/npm-ethereumjs-abi-0.6.8/node_modules/ethereumjs-abi" as pattern ["ethereumjs-abi@0.6.8","ethereumjs-abi@^0.6.8","ethereumjs-abi@^0.6.8"]. This could result in non-deterministic behavior, skipping.
error An unexpected error occurred: "Commit hash required".

How do I solve the error 'Commit hash required"?

### LINKED ISSUE #139: [SUGGESTION] Error running JSON-RPC server: The number NaN cannot be converted to a BigInt because it is not an integer
https://github.com/smartcontractkit/hardhat-starter-kit/blob/ca2bd3d1de672005c8db0fe747c659724e174a32/hardhat.config.ts#L24

This line is causing the next error when running `yarn hardhat node`:
`Error running JSON-RPC server: The number NaN cannot be converted to a BigInt because it is not an integer`

A possible solution would be to set a default value for the const FORKING_BLOCK_NUMBER or adding FORKING_BLOCK_NUMBER to .env.example so devs knows that it's a mandatory env.

### LINKED ISSUE #141: Upgrade to the hardhat toolbox
`@nomicfoundation/hardhat-toolbox` has a lot of what we need installed, and makes our config file a little smaller and neater.

### PR DESCRIPTION
**Pull request by andrejrakic**
## Description

- Migrated project from `hardhat-waffle` to `hardhat-toolbox`
- Removed the `hardhat-deploy` plugin and refactored some parts of the code to remove as many third-party dependencies as possible. Currently `package.json` file consists of only four devDependencies packages: `@chainlink/contracts`, `hardhat`, `@nomicfoundation/hardhat-toolbox` & `solhint`. 
- Extended the content of `.env.example` file.
- Added option for `vrfSubscriptionId` param to be both in network config file and `.env` file.
- Added smart contract examples for NFT Floor Prices and Proof Of Reserve.

## Related issues and PRs
- Fixes #141
- Fixes #112
- Closes #123
- Fixes #136
- Fixes #115 
- Closes #137
- Closes #138
- Fixes #139
- Fixes #89

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -5,4 +5,6 @@ ALCHEMY_MAINNET_RPC_URL="https://eth-mainnet.alchemyapi.io/v2/your-api-key"
 REPORT_GAS=true
 COINMARKETCAP_API_KEY="YOUR_KEY"
 AUTO_FUND=true
-VRF_SUBSCRIPTION_ID=YOUR_SUBSCRIPTION_ID
\ No newline at end of file
+VRF_SUBSCRIPTION_ID=YOUR_SUBSCRIPTION_ID
```
Comment (PatrickAlphaC) [inline .env.example:8]: I personally prefer this in the `hardhat.config.js` but if you like it here that's cool.

File: .husky/pre-commit [hunk 2]
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
+. "$(dirname "$0")/_/husky.sh"
+
+npx lint-staged
```
Comment (PatrickAlphaC) [inline .husky/pre-commit:4]: I don't see pre-commit in the readme. Is this something people should install?

File: README.md [hunk 3]
```diff
@@ -74,24 +74,29 @@ cd hardhat-starter-kit
 ```
 Then:
 ```
-yarn
+npm install
 ```
 
-or
+The recommendation is to use npm 7 or later. If you are using an older version of npm, you'll also need to install all the packages used by the toolbox.
```
Comment (PatrickAlphaC) [inline README.md:80]: Is yarn not cool anymore?

File: README.md [hunk 4]
```diff
@@ -110,7 +115,7 @@ If you run `yarn hardhat --help` you'll get an output of all the tasks you can r
 ## Deploying Contracts
 
 ```
-yarn hardhat deploy
+yarn deploy
```
Comment (PatrickAlphaC) [inline README.md:118]: Should we update everything to be npm then if we are telling them to install with npm?

File: README.md [hunk 5]
```diff
@@ -200,7 +205,7 @@ yarn hardhat deploy --network goerli
 
 To run staging testnet tests
 ```
-yarn hardhat test --network goerli
+yarn hardhat test-staging
```
Comment (PatrickAlphaC) [inline README.md:208]: `npm test-staging`?

File: contracts/AutomationCounter.sol [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+// ./interfaces/AutomationCompatibleInterface.sol
+import "@chainlink/contracts/src/v0.8/AutomationCompatible.sol";
+
+/**
+ * THIS IS AN EXAMPLE CONTRACT THAT USES HARDCODED VALUES FOR CLARITY.
+ * THIS IS AN EXAMPLE CONTRACT THAT USES UN-AUDITED CODE.
+ * DO NOT USE THIS CODE IN PRODUCTION.
+ */
+
+contract Counter is AutomationCompatibleInterface {
```
Comment (PatrickAlphaC) [inline contracts/AutomationCounter.sol:14]: Change name to `AutomationCounter`?

File: contracts/NFTFloorPriceConsumerV3.sol [hunk 7]
```diff
@@ -0,0 +1,34 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.7;
+
+import "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";
+
+contract NFTFloorPriceConsumerV3 {
```
Comment (PatrickAlphaC) [inline contracts/NFTFloorPriceConsumerV3.sol:6]: Do we need this in here? It's just another price feed. Have people been asking for this?

File: contracts/ProofOfReserveConsumerV3.sol [hunk 8]
```diff
@@ -0,0 +1,34 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.7;
+
+import "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";
+
+contract ReserveConsumerV3 {
```
Comment (PatrickAlphaC) [inline contracts/ProofOfReserveConsumerV3.sol:6]: Same with this, it's the same code as a price feed?

**Review by PatrickAlphaC — COMMENTED** [review]
Left some notes.

We don't want hardhat-deploy anymore?

File: .husky/pre-commit (hunk 2 above)
Comment (aelmanaa) [reply .husky/pre-commit:4]: `npx lint-staged`  : cf. [lint-staged](https://github.com/okonet/lint-staged) which is added as a dependency: https://github.com/smartcontractkit/hardhat-starter-kit/blob/fd8f7b8563bae016c00d0048e31f43c9bc88d2a4/package.json#L33

and pre-commits are managed by husky , which is also added as a dependency: 
https://github.com/smartcontractkit/hardhat-starter-kit/blob/fd8f7b8563bae016c00d0048e31f43c9bc88d2a4/package.json#L32

File: README.md (hunk 4 above)
Comment (PatrickAlphaC) [reply README.md:118]: Well, npm and npx anyways

**Review by PatrickAlphaC — APPROVED** [review]
I think these look good! Will do another run of stuff soon to test it out.

**Comment by ggonzalez94** [comment]
Hi! Any particular reason why hardhat deploy was removed? It is pretty useful in my opinion and even recommended in [Patrick's course](https://youtu.be/gyMwXuJrbJQ)

**Comment by PatrickAlphaC** [comment]
@ggonzalez94 hardhat deploy is more opinionated. I also like hardhat deploy, but you can 100% run without it.

**Pull request merged**